### ISSUE TITLE
Audit blocks for contentOnly mode support

### ISSUE THREAD

**Issue by talldan** [WordPress/gutenberg#65778]
## Description

`contentOnly` mode is becoming more prevalent in the editor after '[Select Mode: Use the content-only behavior in select mode](https://github.com/WordPress/gutenberg/pull/65204#top)' was merged.

Lots of block types still aren't working perfectly in `contentOnly` mode, so this list intends to capture those issues so that they can be worked on in parallel by multiple devs. Lots of the blocks also share similar issues, so hopefully there are higher-level solutions to the problems that can be found (lets discuss these in the comments).

Feel free to add more to this issue if you discover untracked issues.

<details>
<summary>Rough snippet that can be run in post editor's console to generate example blocks in contentOnly locked groups, YMMV 🤷 </summary>

```js
const blockExamples = wp.data.select( 'core/blocks' ).getBlockTypes()
  .filter( ( { example } ) => example )
  .map( ( { name, title, example: { attributes, innerBlocks } } ) => ( { name, title, attributes, innerBlocks } ) );
const createInnerExamples = ( innerBlocks ) => innerBlocks.map( ( { name, attributes, innerBlocks } ) => {
  return wp.blocks.createBlock( name, attributes, createInnerExamples( innerBlocks ?? [] ) );
});
const createExamples = ( examples ) => examples.map( ( { name, title, attributes, innerBlocks } ) => {
  const innerBlockInstances = createInnerExamples( innerBlocks ?? [] );

  return wp.blocks.createBlock( 'core/group', {
    templateLock: 'contentOnly',
    metadata: { name: title }
  }, [ wp.blocks.createBlock( name, attributes, innerBlockInstances ) ]);
});
const innerBlockInstances = createExamples( blockExamples );
const blocks = wp.data.select( 'core/block-editor' ).getBlocks();

wp.data.dispatch( 'core/block-editor' ).insertBlocks( innerBlockInstances, blocks.length );

```
</details>

### All blocks
- [x] [Restrict formatting controls in write mode (contentOnly)](https://github.com/WordPress/gutenberg/pull/71058#top)
- [x] Disable block transforms in write mode - All transform actions (row, stack, grid, etc on the Group block for example) should be disabled for all blocks when the editor is in write mode.
- [ ] Remove tools from third party blocks by default and consider surfacing a way for block toolbar controls to opt into rendering during write mode. 

### Buttons

- [x] User cannot add/remove individual buttons (see #52018)
- [x] ~"~Add before" and "add after" options available in the Button block but they don't do anything~

### Cover

- [x] Background image cannot be modified (should it be possible to modify this?)
- [x] Initial block placeholder can't be interacted with in `contentOnly` mode (possibly the same issue as above).
- [ ] Text inside Cover not always editable. Reproduced intermittently, with Cover containing Groups with text inside.

### Details

PR: #70124

- [x] Details block appears as a single 'Paragraph' in the Inspector Content panel

### File

PR: https://github.com/WordPress/gutenberg/pull/65787

- [x] Block is completely non-interactive, including the placeholder

### Gallery

- [x] Placeholder is non-interactive in contentOnly mode.
- [x] Users cannot add/remove items (should they be able to?) (see #52018)

### Image

- [x] Image inserted while in write mode or pre-inserted at root level  has "add text over image" option available but doesn't have "caption" option.
- [x] Image not having ability to add caption is inconsistent with Audio and Video blocks which do allow adding a caption.
- [ ] For pre-inserted Images with captions, the caption isn't always editable (I've reproduced the bug fairly consistently when the block toolbar is at the bottom of the block).
- [ ] Update the order of the tools in the image block to: 1. Add link, 2. Add caption, 3. Crop, 4. Add text over image, 5. Add duotone. In Write Mode the only one of these available should be "link".
- [ ] Move "Replace" to before the other tools as it's the most critical
- [x] [Hide Cut/Copy commands in Write Mode (contentOnly)](https://github.com/WordPress/gutenberg/pull/71063#top)

### List

PR: https://github.com/WordPress/gutenberg/pull/65836

- [x] Users cannot add/remove list items from the list block or nest list items (how should nested list items appear in the Inspector content panel?) (see #52018)
- [ ] List Items currently have a special case in the code to prevent them appearing in the Inspector Content panel (a more general solution to this should be found as other container blocks will have the same kind of problem for their inner blocks): https://github.com/WordPress/gutenberg/blob/df27bd934838d56271224c76d54402886b88160c/packages/block-editor/src/components/block-inspector/index.js#L242
- [x] "Add before" and "add after" options available in List item but they don't do anything

### Media / Text

- [x] Additional paragraphs can't be added in the content area of the block (similar to Quote)
- [ ] In the Inspector panel, the paragraph that's inside the Media Text block appears after it (again, similar to Quote)

### More

PR: #70100

- [x] User cannot edit 'Read more' text

### Navigation

- [x] Cannot be edited (see https://github.com/WordPress/gutenberg/issues/65699)
- [x] [Add View button to navigation link blocks](https://github.com/WordPress/gutenberg/pull/70986#top)

### Post Title / Featured Image / Excerpt

- [x] Blocks cannot be edited (they have no attributes suitable for the `role: content` declaration)

### Query

- [x] https://github.com/WordPress/gutenberg/issues/66614

### Quote

- [ ] The paragraph that's inside the Quote appears after the Quote block in the Inspector Content panel (to repro, add a quote block inside a group, apply `contentOnly` templateLock to the group, and select the group.)
- [x] Users cannot add more paragraphs (should they be able to? If so should they also be able to add other types of blocks? How do these appear in the Inspector Content panel?) (see #52018)

### Site Logo / Site Title / Site Tagline

PR: https://github.com/WordPress/gutenberg/pull/71627 (makes site tagline a content block)

- [x] Site Tagline cannot be edited (has no attributes suitable for the `role: content` declaration)
- [x] [Fix: Hide heading level and alignment controls in site-title block when in contentOnly mode](https://github.com/WordPress/gutenberg/pull/71057#top)
- [x] [Fix: Hide crop and dimensions controls in site-logo block when in Write Mode](https://github.com/WordPress/gutenberg/pull/71061#top)

### Social Link(s)

- [x] https://github.com/WordPress/gutenberg/pull/66622
- [ ] Allow editing individual link text. Should be button be refactored to use `<RichText />`?
- [ ] Should social link block styles be shown in Write mode for each link or only the social links block container? 
- [x] Additional blocks cannot be added to Social Links (see #52018)
- [x] Social links block can be inserted into a page in Write mode, but it can't be edit. Should it be removed from the inserter?

### Spacer

- [x] Should spacer be resizable?

### Synced Patterns in Write Mode

PR: #67026

- [x] https://github.com/WordPress/gutenberg/issues/66424

### Table

PR: #70114 

- [x] Block is completely non-editable in `contentOnly` mode, including the initial placeholder

### Template Parts

- [x] Template parts are unselectable in write mode, as they're not considered 'sections'. (see https://github.com/WordPress/gutenberg/issues/66463)

**Comment by youknowriad** [comment]
For blocks like "list" and "quote", their "inner blocks area" should be marked as "content" somehow it seems. But how far do we take this: is "cover" block's inner blocks also "content"? I guess not but the line is very blurry, in some situations I would love to be able to add paragraphs within "already designed cover blocks" for instance.

**Comment by talldan** [comment]
> For blocks like "list" and "quote", their "inner blocks area" should be marked as "content" somehow it seems. But how far do we take this: is "cover" block's inner blocks also "content"? I guess not but the line is very blurry, in some situations I would love to be able to add paragraphs within "already designed cover blocks" for instance.

Yep, it does feel a bit blurry. Also whether only paragraphs can be added to Quote/Cover or other types of content blocks too. 

Heading and List might be other basic text blocks that are candidates. Then List is itself a container, so this is where the flat content structure in contentOnly mode starts to become more challenging to represent.

**Comment by richtabor** [comment]
> For blocks like "list" and "quote", their "inner blocks area" should be marked as "content" somehow it seems. But how far do we take this: is "cover" block's inner blocks also "content"?

The most straightforward move is to solve for repeatable inner blocks, like Button within Buttons block, and List Item within the List block first. Where if you press enter, another of the same block is inserted.

**Comment by talldan** [comment]
File block was an easy one to fix so I've made a PR - https://github.com/WordPress/gutenberg/pull/65787.

I've also been experimenting with the List block with mixed results. I'll try to get a PR up soon.

**Comment by richtabor** [comment]
> For blocks like "list" and "quote", their "inner blocks area" should be marked as "content" somehow it seems.

Related https://github.com/WordPress/gutenberg/issues/57911

**Comment by annezazu** [comment]
I just tested Write/Design mode with a few themes and want to underscore the following as we need these:

- Site Title/Site Logo: for themes that have a "link in bio" feel, this completely breaks the experience as you can't edit much of the main pieces you would want to.
- Navigation: for any theme, I'd want to be able to edit navigation. 

**Blocks I think should be on the list:**

Social Icons: it's odd that you can't edit these, especially if you think about a link in bio site

https://github.com/user-attachments/assets/0d4b957a-00fa-46bf-9606-b2dbef274db4

Query Loop block: strange experience where you can edit the "no results" text but nothing else. I think we should expand what's available in block settings in line with what's described here so there's a more visual and clearer action to take: https://github.com/WordPress/gutenberg/issues/63497

https://github.com/user-attachments/assets/e55b8961-64ce-4f7c-a146-669f59bdebf9

**Comment by annezazu** [comment]
Just did user testing with a friend and want to underscore featured image block too. They couldn't select it and, as a result, didn't know what it was or where it would show up! Here's a quick video demoing how awkward this is, audio included: 

https://github.com/user-attachments/assets/18d76431-f342-4be1-ad6f-ef557552884e

Finally, they also ran into issues with the site tagline where it says "write your site tagline" prompt yet they can't select it. Pretty confusing!

**Comment by richtabor** [comment]
> Blocks cannot be edited (they have no attributes suitable for the role: content declaration)

Is there not a way to imitate it? Thinking of site title and tagline here.

**Comment by talldan** [comment]
> Is there not a way to imitate it? Thinking of site title and tagline here.

I think an option might be to switch those blocks over to use block bindings (then the blocks would have a `content` attribute used for the binding).

I'll do an investigation to see what that would require.

**Comment by fabiankaegy** [comment]
@talldan @richtabor I'm not sure I understand how `role: content` would help for the usecase @annezazu is describing here. We can't use `role: content` as a metric to determine whether a locked block that lives within a template should remain to be editable 🤔 Because `role: content` tells you nothing about where that attribute is getting stored 🤔 And since those blocks live inside a template they cannot have any actual attributes updated since it would affect every single post and not just the current one. So the block needs to tell the editor in some other way which fields it saves to a separate non attributes field so those can be enabled. Today that happens via a filterable array containing all the blocks that should remain editable when displayed in a locked template.

**Comment by talldan** [comment]
> We can't use role: content as a metric to determine whether a locked block that lives within a template should remain to be editable 🤔 Because role: content tells you nothing about where that attribute is getting stored

I think it can be used as a metric most of the time for `contentOnly` locking, where the attribute definition does determine whether the block is editable.

> Because role: content tells you nothing about where that attribute is getting stored 🤔 And since those blocks live inside a template they cannot have any actual attributes updated since it would affect every single post and not just the current one.

The attribute would have a binding to the site title, so when editing it, the site title would be updated the same way it is today. There shouldn't be any change in functionality. The reason it'd be beneficial is that the block would have a more complete definition and the `content` attribute could be marked with `role: content`.

There's a PR available at #67260 for testing. I might try a different approach, but I think the PR shows the general idea works.

> So the block needs to tell the editor in some other way which fields it saves to a separate non attributes field so those can be enabled. Today that happens via a filterable array containing all the blocks that should remain editable when displayed in a locked template.

If we're talking about page editing in the site editor, I think this is a different concern, though maybe closely connected. When editing in this way, the container blocks that are part of the template are `disabled` (so contentOnly editing is inactive), and then there are those specific blocks you mention that remain editable in that mode (post title, featured image). That probably won't change (site title would still be disabled), though maybe with a binding it's possible to determine where the data is coming from, so it could be possible to revisit the hard-coded array and instead check which parts of the template are bound to the post.

**Comment by tellthemachines** [comment]
I'm noticing differences in the behaviour of some blocks depending on whether they're at root level or inside a container.

* Details block: can't edit summary only if it's inside a container, can edit if it's at root level. 
* Quote and Media & Text blocks: can't add new Paragraphs on Enter if they're at root level, but can if they're inside another container; however trying to create different block with the slash inserter produces an error.
* Image block: can't add or remove captions. If the block is at root level "Add text over image" is available, but not if it's inside a container.

**Comment by youknowriad** [comment]
> Quote and Media & Text blocks: can't add new Paragraphs on Enter if they're at root level, but can if they're inside another container; however trying to create different block with the slash inserter produces an error.

This sounds like a recent regression for me. AFAIK conceptually for the moment, it shouldn't be possible to add blocks even within containers. (maybe this regression impacts the two other cases, but that I don't know for sure)

**Comment by aaronrobertshaw** [comment]
While testing blocks wrapped in contentOnly locked groups, I noticed some weird behaviour where double-clicking would select multiple blocks at the top level. I could also replicate the same behaviour within the site editor. I'm out of time to dig into this but I can revisit on Monday when I'm back online.

| Post Editor | Site Editor |
|---|---|
| <video src="https://github.com/user-attachments/assets/e99b599f-acc2-4540-8446-f2b0585a03cd" /> | <video src="https://github.com/user-attachments/assets/e6a6c053-6b1c-42f2-a040-38d884be2bc9" /> |

**Comment by ramonjd** [comment]
> Allow editing individual link text. Should be button be refactored to use <RichText />?

I added this question to the social link block. There's a bit of a disconnect between Write and Design mode ([context](https://github.com/WordPress/gutenberg/pull/66622/files#r1833999347)) for this block, but a general question occurred to me:

How should blocks know about write mode? Many blocks are using the [useBlockEditingMode](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-editing-mode/index.js) hook  to do custom things based on the "Write/Design" mode.

**Comment by fabiankaegy** [comment]
> How should blocks know about write mode? Many blocks are using the [useBlockEditingMode](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-editing-mode/index.js) hook to do custom things based on the "Write/Design" mode.

As far as I understand things the `write` design mode simply sets blockEditingMode of every single block to `contentOnly` (with some more nuance. Like some also get disabled etc). But it should be possible to simply rely on `useBlockEditingMode()` like we do today to understand if you are in design or in writing mode

**Comment by ramonjd** [comment]
I noticed something while testing https://github.com/WordPress/gutenberg/pull/67372 - should nested template parts be draggable? 

<img width="676" alt="Screenshot 2024-12-18 at 5 22 43 pm" src="https://github.com/user-attachments/assets/d2560151-0eac-4331-a2b7-2e4df7b2a850" />

When editing a template, there seems to be a general inconsistency with write mode blocks draggability - image blocks are draggable, but paragraphs aren't.

Should anything be draggable in write mode?

**Comment by talldan** [comment]
> Should anything be draggable in write mode?

Only blocks that are sections (direct children of the section root). In your screenshot it might be correct if the top level 'Group' is the section root (it has the `<main>` tag).

> image blocks are draggable, but paragraphs aren't.

This part might be a bug, recorded in #67408. Or it might be that the image was also a direct child of the section root.

**Comment by t-hamano** [comment]
There are efforts underway to add `"role": "content"` to block attributes.

I'm wondering if it's appropriate to add `"role": "content"`  to **dynamic blocks**. Does anyone know anything about this?

 In #67260, it looks like there were technical obstacles to applying `contentOnly` to dynamic blocks.

If `"role": "content"` is not suitable for dynamic blocks at this time, we will need to revert the following two merged PRs:

- https://github.com/WordPress/gutenberg/pull/70100
- https://github.com/WordPress/gutenberg/pull/70148

cc @Mamaduka @Sukhendu2002 @yogeshbhutkar @SainathPoojary @Infinite-Null @himanshupathak95  I think you've been working on this task recently

**Comment by Mamaduka** [comment]
@talldan, please correct me if I'm wrong, but I think this tracking issue is for making the "content" fields editable when a parent has the `contentOnly` template lock.

Yes, block binding seems to have a limitation, but that seems like a separate issue. The allowed list of block types that can use bindings is hardcoded at the moment.

https://github.com/WordPress/gutenberg/blob/839a22733e77cca96c53eb690a4a1840a7aa013f/packages/block-editor/src/utils/block-bindings.js#L14-L19

**Comment by talldan** [comment]
> @talldan, please correct me if I'm wrong, but I think this tracking issue is for making the "content" fields editable when a parent has the contentOnly template lock.

Sorry for the late reply. Yes, it's about `contentOnly` mode.

I think `role: content` is ok for dynamic blocks. My understanding is that it only affects blocks in the editor.

#67260 was an exploration to solve the issue for blocks that don't have attributes for their content (e.g. post title uses the value from the post 'entity' and doesn't store the title in an attribute), so there's no way to add `role: content` on an attribute. My idea was to switch those blocks to use bindings so that they'll have a bound `content` attribute that can also be marked as `role: content`.

**Comment by andrewserong** [comment]
A couple of general questions about how the Image block should work:

* We currently show the cover controls in write mode (add text over image), which seems like a destructive action. Should we prevent that in write mode? I'm leaning toward preventing it because once you've converted the image to Cover, we don't have controls for converting it back
* Similarly, should adding or removing captions be allowed in write mode? (Whichever way we go here should be consistent across images, and videos, too)

**Comment by richtabor** [comment]
> We currently show the cover controls in write mode (add text over image), which seems like a destructive action. She we prevent that in write mode? I'm leaning toward preventing it because once you've converted the image to Cover, we don't have controls for converting it back

I'm not aware we support cover controls, other than "Replace" for cover blocks in write mode: 

<img width="3754" height="3526" alt="Image" src="https://github.com/user-attachments/assets/aa391847-d4d9-46f0-9c8a-3d92eab6bcb1" />

> Similarly, should adding or removing captions be allowed in write mode? (Whichever way we go here should be consistent across images, and videos, too)

Write mode is for editing existing content, not introducing new material (including captions imo).

**Comment by tellthemachines** [comment]
I have a PR #71232  to fix some of the issues with being able (or not) to add blocks inside other blocks in write mode. It assumes that if a block has a limited list of allowed blocks _and_ has a content role, we should be able to insert new blocks inside it. This avoids having to create a new mode or attribute to solve the problem, but it does mean that we need to look at which blocks might need a "content" role (e.g. Buttons doesn't have one)

I'd love some feedback on this approach.

One problem that we're potentially facing is if we want to be able to add more blocks inside a Quote or a Details block. They allow adding all types of blocks in them, but that doesn't make sense in write mode. How should we restrict the editing of these blocks? Not allow adding anything, or allow adding just Paragraphs, or...?

Also, it would be great to clarify whether we _should_ have the ability to insert new blocks at root level in write mode (which we currently do, I suspect because this mode is so closely coupled with zoom out mode). The experience is broken, because we're allowed to insert a Group block that then stays invisible/can't be interacted with. This also extends to the ability to duplicate or delete existing blocks. 

How similar an experience should write mode be to zoom out mode? I'm not convinced they should have the same affordances, as one exists fundamentally to rearrange sections and the other to update text and other content.

**Comment by andrewserong** [comment]
> Write mode is for editing existing content, not introducing new material (including captions imo).

Thanks for confirming, there should be a few simple(ish) PRs we can add to make the behaviour consistent for the Image block. To make a start I've opened https://github.com/WordPress/gutenberg/pull/71329 to hide the "convert to cover" controls in the Image block when in contentOnly mode.

**Comment by andrewserong** [comment]
> One problem that we're potentially facing is if we want to be able to add more blocks inside a Quote or a Details block. They allow adding all types of blocks in them, but that doesn't make sense in write mode. How should we restrict the editing of these blocks? Not allow adding anything, or allow adding just Paragraphs, or...?

That's a good question — I feel like for a simple Quote block, maybe offering paragraphs is the simplest way to go, as it's the default block, so it's fairly innocuous. I wonder how folks tend to be using the Details block? If it's for a quick aside or explanation, then maybe Paragraph suffices, too 🤔. On the other hand, if someone's writing content, perhaps they'd want the Details block to have the same access to blocks as whatever is directly above the Details block?

> Also, it would be great to clarify whether we should have the ability to insert new blocks at root level in write mode (which we currently do, I suspect because this mode is so closely coupled with zoom out mode). The experience is broken, because we're allowed to insert a Group block that then stays invisible/can't be interacted with. This also extends to the ability to duplicate or delete existing blocks.

This one's interesting, because we _do_ want to be able to insert new blocks at the root level while editing a post or page (writing content), and of course insert patterns at the root of the main Group block in the site editor. I wonder if it's that we shouldn't be able to insert an empty Group block?

> How similar an experience should write mode be to zoom out mode? I'm not convinced they should have the same affordances, as one exists fundamentally to rearrange sections and the other to update text and other content.

That's another good question — I'd assumed that while zoomed out you might want to quickly update a heading, swap out an image, that kind of thing, so I thought it was intentional for there to be some overlap there?

**Comment by richtabor** [comment]
> One problem that we're potentially facing is if we want to be able to add more blocks inside a Quote or a Details block. They allow adding all types of blocks in them, but that doesn't make sense in write mode. How should we restrict the editing of these blocks? Not allow adding anything, or allow adding just Paragraphs, or...?

I think of content as content. If you can edit a quote block and it's content "attribute" allows for receiving many blocks, I should be able to add those blocks that also have editable contents. I wouldnt think that content roles behave differently, regardless of how a block is architected (a true content role/attribute vs. innerblocks). 

> How similar an experience should write mode be to zoom out mode? I'm not convinced they should have the same affordances, as one exists fundamentally to rearrange sections and the other to update text and other content.

I see zoom out as the birds-eye-view, not the in-the-weeds / editing view. Zoom out should make it easy to rearrange/remove/add/apply styles to patterns, at that top-level pattern-level. I would expect to zoom in to make block-level edits.

> This one's interesting, because we do want to be able to insert new blocks at the root level while editing a post or page (writing content), and of course insert patterns at the root of the main Group block in the site editor. I wonder if it's that we shouldn't be able to insert an empty Group block?

I don't see "Write" mode (which likely should be renamed to something more purposeful/clear) as "this is the best way to write". The current naming convention is not great. It's more of a simpler/streamlined editing experience/edit mode for content that already exists (however so). I would not expect posts or pages to be in Write mode out of the gate, but likely a mode that is engaged for them, or once a page has been built-out (designed with full design mode).

> The experience is broken, because we're allowed to insert a Group block that then stays invisible/can't be interacted with. This also extends to the ability to duplicate or delete existing blocks.

I think we should consider being able to insert content role blocks. If we can add, we should be able to delete. Worth exploring these.

**Comment by ramonjd** [comment]
> "Add before" and "add after" options available in the Button block but they don't do anything

I can no longer reproduce this on trunk in pages/templates. I'll tick it off for now unless folks see something I'm not:

<img width="860" height="149" alt="Image" src="https://github.com/user-attachments/assets/ed7363a7-d13d-4cef-a064-dd532834471b" />

**Comment by tellthemachines** [comment]
> I can no longer reproduce this on trunk in pages/templates.

Yeah that was fixed in #71232 as a consequence of [fixing the conditions for `canInsertBlock`](https://github.com/WordPress/gutenberg/pull/71232/files#diff-a5be080e3d537784f72ea8582ae6642b6e71d2b7989562bfd4c11cb3983a8193R49).

**Comment by tellthemachines** [comment]
I just went through the audit and checked the items that were fixed by https://github.com/WordPress/gutenberg/pull/71232.

Most issues around blocks that we'd like to be able to add to in `contentOnly` mode were solved by that PR, except for Buttons and Social icons blocks (which don't have content roles). Once we've fixed things for those two blocks we should be able to close #52018, correct? Cc @fabiankaegy 

Edit: I just created #71539 to try solving Buttons and Social

**Comment by tellthemachines** [comment]
I just marked complete:

* Buttons and Social items that were fixed in #71539 
* "Post Title / Featured Image / Excerpt" because in testing they all work in write mode (I'm struggling to get Excerpt to work in a post or page template, but it works well inside the Query block). They all have `"role": "content"` in one or more of their attributes.

**Comment by talldan** [comment]
I also marked Site Tagline complete after https://github.com/WordPress/gutenberg/pull/71627 was merged.

**Comment by t-hamano** [comment]
If I understand correctly, we're aiming to stabilize write mode in 6.9 release, so let's add this issue to the Editor tasks borad.

**Comment by intelligence** [comment]
Not sure if this is the correct issue to comment on. But just found that I cannot add or rearrange images added to a gallery. My use case is that I've built a custom block consisting of a template with a heading and a gallery and applied contentOnly with the hopes that the editor could edit the heading and add/remove/rearrange images in the gallery, but it's not possible.

Is there some work around?

**Comment by tellthemachines** [comment]
@intelligence I'm struggling to reproduce that both by adding "templateLock": "contentOnly" to a block and by editing a pattern containing a gallery. What does your custom block markup look like, and how are you applying contentOnly to it?

**Comment by intelligence** [comment]
Hey @tellthemachines, thanks for taking a look!

Here's the code for my custom block:

```
import { useBlockProps, InnerBlocks } from '@wordpress/block-editor';

const TEMPLATE = [
    ['core/heading', { }],
    ['core/gallery', {
        columns: 4,
        randomOrder: false,
        imageCrop: false,
    }],
];

const TEMPLATE_LOCK = 'contentOnly'; 

export default function Edit() {
    return (
        <div { ...useBlockProps() }>
            <InnerBlocks
                template={ TEMPLATE }
                templateLock={ TEMPLATE_LOCK }
            />
        </div>
    );
}
```

Me trying to change order of stuff:

https://github.com/user-attachments/assets/eb9a74b0-c39e-4f3a-93a6-3925b48203a3

This is on WP 6.9.

**Comment by tellthemachines** [comment]
Thanks @intelligence I can reproduce the issue with the provided block in 6.9. The good news is that it's not reproducible in Gutenberg, so it has since been fixed. I don't believe there's a workaround for 6.9 but you can enable that functionality by using the Gutenberg plugin. Otherwise, if you're happy to wait for 7.0 the fix will be shipping in core then!

**Comment by intelligence** [comment]
That's great! Sorry that I didn't try the Gutenberg plugin before coming here. Installed it and can confirm it works like intended.
